Create the HookPriorityLimit Sniff

antonioeatgoat opened the pull request:
It checks if `PHP_INT_MAX` or `PHP_INT_MIN` are added used for `add_filter` or `add_action`.

**Please check if the PR fulfills these requirements**
- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes/features)
- [X] Docs have been added/updated (for bug fixes/features)


**What kind of change does this PR introduce?**
Feature


**What is the current behavior?**
No warnings are raised for `PHP_INT_MAX` or `PHP_INT_MIN`.


**What is the new behavior (if this is a feature change)?**
It raises a warning if:

- `PHP_INT_MAX` is used for `add_filter`.
- `PHP_INT_MIN` is used for `add_action` or `add_filter`.


**Does this PR introduce a breaking change?** 
No


**Other information**:
The code is not very robust I think, but I spent a a good amount of time on it and this is the only working solution I was able to produce, feel free to discard it totally if there is a better way.

Anyway, there are many things I am still not sure:
- The Sniff name.
- The error code (The code itself and if they should be the same or not in both case).
- The error messages.

antonioeatgoat committed:
Create the HookPriorityLimit Sniff

It checks if `PHP_INT_MAX` or `PHP_INT_MIN` are added used for `add_filter` or `add_action`.

antonioeatgoat committed:
Convert errors to warnings

antonioeatgoat committed:
Update documentation

tfrommen commented on Inpsyde/Sniffs/CodeQuality/HookPriorityLimitSniff.php:
```diff
@@ -0,0 +1,101 @@
+use PHP_CodeSniffer\Files\File;
+use PHP_CodeSniffer\Sniffs\Sniff;
+
+class HookPriorityLimitSniff implements Sniff
```
Can we maybe name this `HookPrioritySniff`, without the "limit"?

antonioeatgoat replied:
This was my first choice actually, later I changed it but yes, I agree that the shorter version is probably better I'll change it.

tfrommen commented on Inpsyde/Sniffs/CodeQuality/HookPriorityLimitSniff.php:
```diff
@@ -0,0 +1,101 @@
+            return;
+        }
+
+        $parameters = $this->stringParameters($phpcsFile, $stackPtr + 1);
```
Do you know about [`PassedParameters`](https://phpcsutils.com/phpdoc/classes/PHPCSUtils-Utils-PassedParameters.html)?

tfrommen commented on Inpsyde/Sniffs/CodeQuality/HookPriorityLimitSniff.php:
```diff
@@ -0,0 +1,101 @@
+
+        $parameters = $this->stringParameters($phpcsFile, $stackPtr + 1);
+
+        if (in_array('PHP_INT_MAX', $parameters, true) && $functionName === 'add_filter') {
```
Why are we checking for **any** presence? We know the priority is the third argument, right?

tfrommen commented on Inpsyde/Sniffs/CodeQuality/HookPriorityLimitSniff.php:
```diff
@@ -0,0 +1,101 @@
+        $parameters = $this->stringParameters($phpcsFile, $stackPtr + 1);
+
+        if (in_array('PHP_INT_MAX', $parameters, true) && $functionName === 'add_filter') {
+            $phpcsFile->addWarning(
```
Do you know about [`MessageHelper`](https://phpcsutils.com/phpdoc/classes/PHPCSUtils-Utils-MessageHelper.html)?

antonioeatgoat replied:
Not sure about what do you expect here to being improved. I think I printed the message the same way we do in every other sniff we have. Am I missing something? 🤔

gmazzap replied:
MessageHelper was not used when I introduced PHPUtils.
The reason is that it is only helpful when you have replacements in the message, otherwise it requires a longer signature without real benefit, IMO.
Anyway, considering it is mnot used anywhere I think it is also fine not using it here.

tfrommen commented on Inpsyde/Sniffs/CodeQuality/HookPriorityLimitSniff.php:
```diff
@@ -0,0 +1,101 @@
+
+        if (in_array('PHP_INT_MAX', $parameters, true) && $functionName === 'add_filter') {
+            $phpcsFile->addWarning(
+                'PHP_INT_MAX applied as filter priority.',
```
To me, this is missing the Why.

Also, "used" sounds more appropriate than "applied".

And if we use "hook priority", which is just like the name of the sniff, we don't have to differentiate or name both filter and action.

How about something like this:

```
Found PHP_INT_MAX used as hook priority. This makes it hard, if not impossible to reliably remove the callback.
```

antonioeatgoat replied:
> Also, "used" sounds more appropriate than "applied".

Thank you for this 👍 

For the message in general, I like the longer and more descriptive version, but I think we never explain so much in other custom sniffs we have. I tried to stick with their errors messages style 🤔

gmazzap replied:
We're not very consistent in this matter. [Some rules have very long explaination](https://github.com/inpsyde/php-coding-standards/blob/development/Inpsyde/Sniffs/CodeQuality/NoAccessorsSniff.php#L91-L93) other much shorter. So if we like @tfrommen message (and I do), I guess we can use it.

antonioeatgoat replied:
Well that's true, that one is quite long 😅 

I just pushed the changed and I've introduced longer error description. Still I used two different messages because I think the reason is a bit different based on the case. Please let me know if it makes sene for you.

tfrommen reviewed (commented):
Thank you for working on this! 👏

I feel this sniff can be simplified and also improved by using PHPCSUtils.

antonioeatgoat committed:
Rename the sniff in `HookPrioritySniff`

antonioeatgoat committed:
Simplify the code

Use the utility method `PassedParameters::getParameter()` in order to get rid of a lot of custom code.

antonioeatgoat committed:
Update the sniff name in the readme
